bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/fixity/FixityChecker.scala

```diff
@@ -50,85 +120,175 @@ trait FixityChecker extends Logging {
 
-      // Checksum does not match that provided
-      case Success(checksum) =>
-        if (checksum != expectedFileFixity.checksum)
-          FileFixityMismatch(
+      case Some(storedValue) if storedValue == fixityTagValue => {
+        Right(
+          FileFixityCorrect(
             expectedFileFixity = expectedFileFixity,
             objectLocation = objectLocation,
-            e = FailedChecksumNoMatch(
-              actual = checksum,
-              expected = expectedFileFixity.checksum
-            )
+            size = size
           )
-        else
-          FileFixityCorrect(
+        )
+      }
+
+      case Some(storedValue) =>
+        Left(
+          FileFixityMismatch(
             expectedFileFixity = expectedFileFixity,
             objectLocation = objectLocation,
-            size = inputStream.length
+            e = new Throwable(
+              s"Checksum mismatch: object tags ($storedValue) does not match manifest checksum (${expectedFileFixity.checksum})"
+            )
           )
+        )
     }
+
+  private def verifyChecksumByReadingObject(
+    expectedFileFixity: ExpectedFileFixity,
+    objectLocation: ObjectLocation,
+    algorithm: HashingAlgorithm,
+    size: Long
+  ): Either[FileFixityError, FileFixityCorrect] =
+    for {
+      inputStream <- getInputStream(objectLocation, expectedFileFixity)
+
+      result <- Checksum.create(inputStream, algorithm) match {
+        // Failure to create a checksum (parsing/algorithm)
+        case Failure(e) =>
+          Left(
+            FileFixityMismatch(
+              expectedFileFixity = expectedFileFixity,
+              objectLocation = objectLocation,
+              e = FailedChecksumCreation(algorithm, e)
+            )
+          )
+
+        // Checksum does not match that provided
+        case Success(checksum) =>
+          if (checksum != expectedFileFixity.checksum) {
+            Left(
+              FileFixityMismatch(
+                expectedFileFixity = expectedFileFixity,
+                objectLocation = objectLocation,
+                e = FailedChecksumNoMatch(
+                  actual = checksum,
+                  expected = expectedFileFixity.checksum
+                )
+              )
+            )
+          }
+
+          // Checksum is correct
+          else {
+            assert(inputStream.length == size)
+            Right(
+              FileFixityCorrect(
+                expectedFileFixity = expectedFileFixity,
+                objectLocation = objectLocation,
+                size = inputStream.length
+              )
+            )
+          }
+      }
+    } yield result
+
+  private def writeChecksumTag(
+    expectedFileFixity: ExpectedFileFixity,
+    objectLocation: ObjectLocation,
+    fixityTags: Map[String, String]): Either[FileFixityCouldNotWriteTag, Unit] =
+      Try {
+        debug(s"Adding tags $fixityTags to $objectLocation")
+        tags
+          .update(objectLocation) { existingTags =>
```

Comment:
I think this is only a weak guarantee though as the tags are not "locked" during this checking operation. I think you probably already know that & I would agree this check is worth doing in any case.